### PR TITLE
adding a checker that warns about workloads which are not covered by …

### PR DESCRIPTION
**Pull request by shireenf-ibm**
…any authorization policy

** Describe the change **
new checker files:

-> I have added a new checker under business/checkers/authorization called "uncovered_workload_checker.go", this checker gets lists of authorizationPolicies and Workloads and checks if there is a workload in the workloadslist that is not covered by any of the given authorization policies.

-> a test file called "uncovered_workload_checker_test.go" was added too, it tests "good" and "bad" cases of the checker.
a comment that describes the testing case was added before each call of the "Check" method of the new checker.

more changes in code are:
-> " business/checkers/authorization_policies_checker.go" file : 
since my checker's main parent is AuthorizationPolicyChecker, I have added the call for the new checker here in the "runGroupChecks" as it checks all auhorizationPolicies.

->  "models/istio_validation.go" : added a description of the new checker, to be compared with in the checker tests.
->  "tests/data/authorization_policy_data.go" : for my tests I needed to create new authorization policies such as: empty authorization policy or an authorization policy with specified namespace and selector. So new creators were added here.

** Issue reference **
Fixes https://github.com/kiali/kiali/issues/4368

UI PR: [#2278](https://github.com/kiali/kiali-ui/pull/2278)
Helm PR: [#120](https://github.com/kiali/helm-charts/pull/120)
Operator PR: [#467](https://github.com/kiali/kiali-operator/pull/467)

** Backwards incompatible? ** 
- Is your change introducing backwards incompatible changes?No

### CODE REVIEW

File: business/checkers/authorization/uncovered_workload_checker.go
Comment (shireenf-ibm) [inline business/checkers/authorization/uncovered_workload_checker.go:57]: Hi,
I have a question here (line 57).
when I want to compare the apNamespace (authorization policy namespace) with the "Istio root namespace".

I wrote this  `apNamespace == "istio-system" ` compare statement .

because the following one  didn't work for me (test failed unexpectedly)
`apNamespace == config.Get().IstioNamespace` 
(note: I used the kiali/config import as used in \business\checkers\peer_authentication_checker.go but still the tests failed unexpectedly) 

I couldn't understand why it failed for me,  so preferred to use the "safe by hand" comparison.

Can it be a "not running on cluster" problem on my workspace?

Comment (shireenf-ibm) [inline business/checkers/authorization/uncovered_workload_checker.go:57]: Hi,
I have a question here (line 57).
when I want to compare the apNamespace (authorization policy namespace) with the "Istio root namespace".

I wrote this apNamespace == "istio-system" compare statement .

because the following one didn't work for me (test failed unexpectedly)
apNamespace == config.Get().IstioNamespace
(note: I used the kiali/config import as used in \business\checkers\peer_authentication_checker.go but still the tests failed unexpectedly)

I couldn't understand why it failed for me, so preferred to use the "safe by hand" comparison.

Can it be a "not running on cluster" problem on my workspace?

Comment (xeviknal) [reply business/checkers/authorization/uncovered_workload_checker.go:57]: Hi @shireenf-ibm! Good you are asking :)
This situation is due to the initialization of the config values. The values are present when you are starting the whole server up. Otherwise, the config file is empty.

In our unit testing suite, kiali doesn't start the server so that is why you found it empty. We tipically do the following:
```go
	conf := config.NewConfig()
	config.Set(conf)
```

You can find one example here:
https://github.com/kiali/kiali/blob/e38167d651d0334e1f75811e1b6536d6beb3f6e5/business/checkers/destinationrules/multi_match_checker_test.go#L16-L17

**Comment by lucasponce** [comment]
Hi @shireenf-ibm, we introduced a large refactoring (https://github.com/kiali/kiali/pull/4389) that would affect to your work.
@xeviknal or I are available and we could help you to update it or resolve any conflict you may have.

Thanks !

File: business/checkers/workloads/uncovered_workload_checker.go
Comment (xeviknal) [inline business/checkers/workloads/uncovered_workload_checker.go:6]: nitpick: we usually split the imports in different blocks:
1. native modules
2. dependencies modules
3. kiali modules

Do you mind splitting this import in two? Thanks!

File: business/checkers/workloads/uncovered_workload_checker_test.go
Comment (xeviknal) [inline business/checkers/workloads/uncovered_workload_checker_test.go:11]: Same thing for the imports here! Sorry for the nitpick.

File: business/istio_validations.go
Comment (xeviknal) [inline business/istio_validations.go:107]: There is one thing missing to be consistent with services though.
Right now, we are returning all the validation even we only need the ones related to the workload.
Do you mind add the else case in the following if? It will cast out validations not pointing to workloads.

**Review by xeviknal — COMMENTED** [review]
@shireenf-ibm we are getting close! thanks for keep putting some time in here.
There are a couple of nitpicks and one comment to change. After that, it will look really good to go.
Thank you again :clap:

File: business/workloads.go
Comment (xeviknal) [inline business/workloads.go:152]: It looks like there was one message that didn't arrive into my last review. Apologies.
There is a problem here. workloadList is still an empty struct at this point. It is actually filled in in the line 160, right at the end of the loop (`workloadList.Workloads = append(workloadList.Workloads, *wItem)`)
If you place this selected line after the loop, where you have the Validation assignation, it should work.

By checking the following endpoints you should see the results:
`/api/namespaces/bookinfo/workloads?validate=true` for Workload List
`/api/namespaces/bookinfo/workloads/details-v1?validate=true` for the details page.

I don't know if you have the local environment fully installed, here you have some more details: https://github.com/kiali/kiali#developer-setup

And if you need any further assistance, I can help. You can find me on the istio slack if you need me.

**Review by xeviknal — CHANGES_REQUESTED** [review]
Looks almost good to me. There is one small problem in my comment. It should take so long to change it.

**Comment by xeviknal** [comment]
@shireenf-ibm good! Thanks for the quick response. I am throwing the CI to see everything is correct :)

Tipically, backend PRs that need frontend aren't merged until the frontend PR is approved and ready to go as well. So I am putting the `requires UI PR` label to visualize that.

**Comment by xeviknal** [comment]
I see there is a problem on the documentation side of things. You need to update the swagger file.
In order to do so, we have in place a makefile that automatically does it.

```
make swagger-install swagger-gen swagger-validate
```

**Review by xeviknal — DISMISSED** [review]
LGTM! Great work @shireenf-ibm !

However don't merge the PR until the frontend is ready

**Comment by lucasponce** [comment]
> However don't merge the PR until the frontend is ready

Hi @shireenf-ibm, just a ping to check how the frontend PR is going, to see if @hhovsepy could review it all together.

@xeviknal is off for some weeks but rest of the team can help you to progress on this work, so don't hesitate to contact us.

Thanks

**Comment by shireenf-ibm** [comment]
> > However don't merge the PR until the frontend is ready
> 
> Hi @shireenf-ibm, just a ping to check how the frontend PR is going, to see if @hhovsepy could review it all together.
> 
> @xeviknal is off for some weeks but rest of the team can help you to progress on this work, so don't hesitate to contact us.
> 
> Thanks

Hi 
actually I was assuming that @xeviknal  is working on the frontend PR , since I am not familiar with the kiali-UI workspace.
So, I would like to get some instructions on how to contribute to kiali-ui and what changes should be done to support this PR in order to progress with this

**Comment by lucasponce** [comment]
> So, I would like to get some instructions on how to contribute to kiali-ui and what changes should be done to support this PR in order to progress with this

Ah, sorry, I missed the discussion then, I cc @leandroberetta that is familiar with the UI code and he can help on this.

File: business/checkers/workloads/uncovered_workload_checker.go
Comment (leandroberetta) [inline business/checkers/workloads/uncovered_workload_checker.go:46]: We've recently added a new feature to correctly reference the admiistrative root namespace in  Istio, similar to "IsIstioNamespace" you have "IsRootNamespace" and in this situation the last should be used. The root namespace defaults to the Istio namespace control plane if not configured in Kiali CR.

**Review by leandroberetta — COMMENTED** [review]
Hi @shireenf-ibm thanks for the contribution.

Only a minor comment about the root namespace, the rest LGTM.

I tested with the UI and I guess we need to take some decisions in how we will show this to the users:

Overview Page:

ap_workloads_overview

As you can see, every namespace with workloads that are not covered by a AP will be add a warning message. Not only the message but the workload is counting as a config object, you can see this in the following debugging session:

image

There is one object key for a Gateway object but also object keys of type workload, I guess this is added since the changes in this PR. For the namespace summary, I guess we can try to filter the workloads, to return to the normal value of "2 config objects analyzed".

The 7 warnings are ok because 7 workloads are not covered by any AP.

A thought that comes to my mind is that this situation will be very common at least in the first steps of anyone using Kiali and also in demos, so I'm not sure if we want to always show this kind of warnings, a solution might be to add a configuration parameter in Kiali CR to be more strict with this kind of situations.

What do you think?.

Workload List:

ap_workloads_list

I guess that in this page, we can add some reference to the warning in the details column, not sure which text "Not covered by Authorization Policy" but it's long for the available space, maybe just a warning icon and a tooltip.

Workload Page

ap_workloads_page

In this section, I was thinking in add a new section called "Insights", because this warning is really an insight we are finding, in fact, the whole feature can be seen as an insight, because validations usually validates Istio Config Objects.

Another example of an insight would be that there is not a global Sidecar affecting to all workloads, but that is a broader discussion and a bigger epic.

So the section would be like:

ap_workloads_insight

With the proper messageo of course.

What do you think about this comments, do you have something different in mind?.

**Comment by lucasponce** [comment]
> In this section, I was thinking in add a new section called "Insights", because this warning is really an insight we are finding, in fact, the whole feature can be seen as an insight, because validations usually validates Istio Config Objects.

Not fully sure if insights can create some confusion. I think this is a similar message as the "missing sidecar" or "missing label" for a workload.

my suggestion would be:
- In the list page, to add the proper message in the "details" column.
- In the workload details page, use same pattern of using an informative icon for this situation.

In both cases I'd add a tooltip with the information detail, but I'd follow same pattern instead to introduce a new card (we can do that in the future if these sets of feature have more evolution, but I don't see it just for this).

I'm +1 with some configuration flag in case that the user don't use the AuthorizationPolicy, as you're right, they will be highly populated with undesired warnings.

**Comment by leandroberetta** [comment]
Hi @shireenf-ibm thanks for adding support for the root namespace.

I was coding the UI part, for now it is draft, you can find it here: https://github.com/kiali/kiali-ui/pull/2278, it would be nice if you provide feedback about how we are displaying the new validation to users.

Also, there is still work to do adding a config flag to disable this validation.

**Comment by shireenf-ibm** [comment]
> Hi @shireenf-ibm thanks for adding support for the root namespace.
> 
> I was coding the UI part, for now it is draft, you can find it here: [kiali/kiali-ui#2278](https://github.com/kiali/kiali-ui/pull/2278), it would be nice if you provide feedback about how we are displaying the new validation to users.
> 
> Also, there is still work to do adding a config flag to disable this validation.

Hi, thanks for the update
it looks very good to me , glad to see how the warning is displayed wishing it'll be helpful for users too

File: business/workloads.go
Comment (leandroberetta) [inline business/workloads.go:162]: @shireenf-ibm as this feature will be optional, and by default this validation will be skipped, we need to add here a call to a function called "StripIgnoredChecks".

For reference, this is done in "runObjectCheckers" at istio_validations.go

**Review by leandroberetta — CHANGES_REQUESTED** [review]
Hi @shireenf-ibm, just a minor fix for allowing this validation to be optional.

**Comment by leandroberetta** [comment]
Hi @shireenf-ibm, thanks for the last update.

I was doing some backend testing applying Authorization Policies and I found the following in my setup:

1. I installed bookinfo application in bookinfo namespace using our hack scripts
2. I created an AuthorizationPolicy that covers the bookinfo namespace workloads

I see that in the workloads list for that namespace there is no validation message but on the workload details, it is, checking the API responses, I see that for:

http://localhost:20001/kiali/api/namespaces/bookinfo/workloads?validate=true

There is no validation errors, but for:

http://localhost:20001/kiali/api/namespaces/bookinfo/workloads?validate=true

The validation message is present

**Comment by shireenf-ibm** [comment]
> Hi @shireenf-ibm, thanks for the last update.
> 
> I was doing some backend testing applying Authorization Policies and I found the following in my setup:
> 
>     1. I installed bookinfo application in bookinfo namespace using our hack scripts
> 
>     2. I created an AuthorizationPolicy that covers the bookinfo namespace workloads
> 
> 
> I see that in the workloads list for that namespace there is no validation message but on the workload details, it is, checking the API responses, I see that for:
> 
> http://localhost:20001/kiali/api/namespaces/bookinfo/workloads?validate=true
> 
> There is no validation errors, but for:
> 
> http://localhost:20001/kiali/api/namespaces/bookinfo/workloads?validate=true
> 
> The validation message is present
I am sorry that happened.
Unfortunately I couldn't run the application on my ws (running with a bash terminal on windows
Can you run this with a debug mode so we can get were the problem is.
Basically when writing the  workloads handler I followed as a reference, the "services" handler. maybe I missed some point on handler
woulf like to get more details on what should be done

**Comment by leandroberetta** [comment]
> > Hi @shireenf-ibm, thanks for the last update.
> > I was doing some backend testing applying Authorization Policies and I found the following in my setup:
> > ```
> > 1. I installed bookinfo application in bookinfo namespace using our hack scripts
> > 
> > 2. I created an AuthorizationPolicy that covers the bookinfo namespace workloads
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I see that in the workloads list for that namespace there is no validation message but on the workload details, it is, checking the API responses, I see that for:
> > http://localhost:20001/kiali/api/namespaces/bookinfo/workloads?validate=true
> > There is no validation errors, but for:
> > http://localhost:20001/kiali/api/namespaces/bookinfo/workloads?validate=true
> > The validation message is present
> 
> I am sorry that happened. Unfortunately I couldn't run the application on my ws (running with a bash terminal on windows Can you run this with a debug mode so we can get were the problem is. Basically when writing the workloads handler I followed as a reference, the "services" handler. maybe I missed some point on handler woulf like to get more details on what should be done

HI @shireenf-ibm, sure, no problem, I'll let you know after debugging!.

**Comment by lucasponce** [comment]
Hi folks,
Do we have an update on this backend PR? 
Is it waiting for a review on this one https://github.com/kiali/kiali-ui/pull/2278?
Thx

**Comment by leandroberetta** [comment]
> Hi folks, Do we have an update on this backend PR? Is it waiting for a review on this one [kiali/kiali-ui#2278](https://github.com/kiali/kiali-ui/pull/2278)? Thx

I need to debug some case where validations are still appearing after ignoring them via CR (Default case).

The UI PR can be reviewed.

I will debug the backend to find the issue.

**Comment by leandroberetta** [comment]
@shireenf-ibm I was debugging a little bit the code and I found I guess the origin of the problem.

When invoking the Workloads list endpoint:

http://localhost:20001/kiali/api/namespaces/bookinfo/workloads?validate=true

Things works well, the validation message is not present as I placed a namespaced auth policy (in this case on bookinfo namespace).

But when invoking an specific workload, its details:

http://localhost:20001/api/namespaces/bookinfo/workloads/kiali-traffic-generator?validate=true

I see the validations are present in the response.

Debugging, I see that when invoking the individual workload details, every workload is being tested against the validation checker:

istio_validations.go (GetValidations)
image

When invoking the workload details, when testing against "hasCoveringAuthPolicy", I see that sometimes returns true and other times false:

image

I guess this is OK with this logic because is testing agains all workloads, from other namespaces that does not have any AuthPolicy. 

The thing is that after all this run checks, there is a MergeValidations, where I guess is the problem, there is a lot of KIA1201 errors that are reduced as an unique error that is place in the response.

So, I guess that this logic need to be reviewed for individual workloads because mainly, it is very expensive to check all workloads and not necessary. Fixing that, it may fix the issue (Because it will only check the required workload).

Let me know if you need more info!

**Comment by hhovsepy** [comment]
In Workloads details page, it shows warning about missing AuthPolicy when actually it exists.
It happens because AuthPolicies are not listed, the criteria needs to be updated to include resources "IncludeIstioResources: true" in methods "fetchWorkloads" and "fetchAllWorkloads" of "istio_validations.go"

**Comment by leandroberetta** [comment]
> In Workloads details page, it shows warning about missing AuthPolicy when actually it exists. It happens because AuthPolicies are not listed, the criteria needs to be updated to include resources "IncludeIstioResources: true" in methods "fetchWorkloads" and "fetchAllWorkloads" of "istio_validations.go"

@hhovsepy thanks, right, I missed that. 

That for sure will fix the problem. The remaining comment I have is that it's not necessary to fetch all workloads when entering into a workload specific details.

**Comment by leandroberetta** [comment]
Hi @shireenf-ibm, as @hhovsepy commented, there is some parameter to list the auth policies in the case I was mentioning (workload details), is ok for you to update this and also do a rebase of the code?.

**Comment by lucasponce** [comment]
@leandroberetta @hhovsepy @shireenf-ibm how do you see this work to merge it in the Sprint 69?

**Comment by hhovsepy** [comment]
` Error: level=warning msg="[runner] Can't run linter unused: buildssa: analysis skipped: errors in package: [/home/runner/work/kiali/kiali/handlers/workloads.go:66:81: cannot use namespace (variable of type string) as context.Context value in argument to business.Validations.GetValidations: missing method Deadline /home/runner/work/kiali/kiali/handlers/workloads.go:66:104: too few arguments in call to business.Validations.GetValidations /home/runner/work/kiali/kiali/handlers/workloads.go:121:81: cannot use namespace (variable of type string) as context.Context value in argument to business.Validations.GetValidations: missing method Deadline /home/runner/work/kiali/kiali/handlers/workloads.go:121:104: too few arguments in call to business.Validations.GetValidations -: could not load export data: no export data for \"github.com/kiali/kiali/handlers\"]"
Error: level=error msg="Running error: buildssa: analysis skipped: errors in package: [/home/runner/work/kiali/kiali/handlers/workloads.go:66:81: cannot use namespace (variable of type string) as context.Context value in argument to business.Validations.GetValidations: missing method Deadline /home/runner/work/kiali/kiali/handlers/workloads.go:66:104: too few arguments in call to business.Validations.GetValidations /home/runner/work/kiali/kiali/handlers/workloads.go:121:81: cannot use namespace (variable of type string) as context.Context value in argument to business.Validations.GetValidations: missing method Deadline /home/runner/work/kiali/kiali/handlers/workloads.go:121:104: too few arguments in call to business.Validations.GetValidations -: could not load export data: no export data for \"github.com/kiali/kiali/handlers\"]"
make: *** [make/Makefile.build.mk:136: lint] Error 3
Error: Process completed with exit code 2.
`

**Comment by shireenf-ibm** [comment]
> `Error: level=warning msg="[runner] Can't run linter unused: buildssa: analysis skipped: errors in package: [/home/runner/work/kiali/kiali/handlers/workloads.go:66:81: cannot use namespace (variable of type string) as context.Context value in argument to business.Validations.GetValidations: missing method Deadline /home/runner/work/kiali/kiali/handlers/workloads.go:66:104: too few arguments in call to business.Validations.GetValidations /home/runner/work/kiali/kiali/handlers/workloads.go:121:81: cannot use namespace (variable of type string) as context.Context value in argument to business.Validations.GetValidations: missing method Deadline /home/runner/work/kiali/kiali/handlers/workloads.go:121:104: too few arguments in call to business.Validations.GetValidations -: could not load export data: no export data for \"github.com/kiali/kiali/handlers\"]" Error: level=error msg="Running error: buildssa: analysis skipped: errors in package: [/home/runner/work/kiali/kiali/handlers/workloads.go:66:81: cannot use namespace (variable of type string) as context.Context value in argument to business.Validations.GetValidations: missing method Deadline /home/runner/work/kiali/kiali/handlers/workloads.go:66:104: too few arguments in call to business.Validations.GetValidations /home/runner/work/kiali/kiali/handlers/workloads.go:121:81: cannot use namespace (variable of type string) as context.Context value in argument to business.Validations.GetValidations: missing method Deadline /home/runner/work/kiali/kiali/handlers/workloads.go:121:104: too few arguments in call to business.Validations.GetValidations -: could not load export data: no export data for \"github.com/kiali/kiali/handlers\"]" make: *** [make/Makefile.build.mk:136: lint] Error 3 Error: Process completed with exit code 2.`

fixed

**Comment by hhovsepy** [comment]
`
Run make lint
golangci-lint run -c ./.github/workflows/config/.golangci.yml --skip-files "doc\.go" --tests --timeout 5m
business/istio_validations_test.go:37: File is not `goimports`-ed with -local github.com/kiali/kiali (goimports)

handlers/namespaces.go:46: File is not `goimports`-ed with -local github.com/kiali/kiali (goimports)

handlers/services.go:139: File is not `goimports`-ed with -local github.com/kiali/kiali (goimports)
      
make: *** [make/Makefile.build.mk:136: lint] Error 1
Error: Process completed with exit code 2.`

**Comment by hhovsepy** [comment]
`Run make lint
golangci-lint run -c ./.github/workflows/config/.golangci.yml --skip-files "doc\.go" --tests --timeout 5m
handlers/namespaces.go:46: File is not `goimports`-ed with -local github.com/kiali/kiali (goimports)

handlers/services.go:139: File is not `goimports`-ed with -local github.com/kiali/kiali (goimports)
      
make: *** [make/Makefile.build.mk:136: lint] Error 1`

**Comment by hhovsepy** [comment]
`
Run make swagger-ci
2022/02/07 15:44:16 
The swagger spec at "./swagger.json" is valid against swagger specification 2.0
kiali_api.md is not correct, remember to run make swagger-gen to update kiali_api.md
make: *** [make/Makefile.build.mk:112: swagger-ci] Error 1
Error: Process completed with exit code 2.`

**Comment by Xunzhuo** [comment]
@shireenf-ibm Have you try to run `make swagger-gen` ?

**Comment by shireenf-ibm** [comment]
> 

image

I am also attaching the kiali_api_copy.md and swagger_copy.json (removed the rm cmd to keep them)
[kiali_api_copy.md](https://github.com/kiali/kiali/files/8022455/kiali_api_copy.md)
[kiali_api.md](https://github.com/kiali/kiali/files/8022456/kiali_api.md)
[swagger_copy.json.txt](https://github.com/kiali/kiali/files/8022458/swagger_copy.json.txt)

**Comment by Xunzhuo** [comment]
Have you git fetch the upstream and update your codes? Maybe there are some updates.

**Comment by hhovsepy** [comment]
@shireenf-ibm you need to run this commands:
`
make swagger-ci
make swagger-gen
`
then commit both kiali_api.md and swagger.json

**Comment by leandroberetta** [comment]
> make swagger-ci make swagger-gen

When the CI problem is fixed, I'm OK as I tested the problem I had before, and now looks good.

**Review by hhovsepy — APPROVED** [review]
Verified now, thank you for contribution @shireenf-ibm

**Comment by skondkar** [comment]
Performed post merge verification

**Pull request merged**
